File: csharp/866.prime-palindrome.cs
/*
 * @lc app=leetcode id=866 lang=csharp
 *
 * [866] Prime Palindrome
 */

using System;

// @lc code=start
public class Solution {

    private static int[] Singles = new int[] {
        2, 2, 3, 5, 5, 7, 7, 11, 11, 11, 11
    };
    
    private static bool IsPrime(int x) {
        if (x <= 2) {
            return x == 2;
        }
        int m = (int)Math.Sqrt(x);
        for (int i = 2; i <= m; ++i) {
            if (x % i == 0) {
                return false;
            }
        }
        return true;
    }
    
    public int PrimePalindrome(int N) {
        if (N <= 11) {
            return Singles[N - 1];
        }

        for (int x = 10; true; ++x) {
            int t = x / 10, y = 0, c = 1;
            while (t > 0) {
                y = y * 10 + t % 10;
                t /= 10; 
                c *= 10;
            }
            y += c * x;
            if (y >= N && IsPrime(y)) {
                return y;
            }
        }
    }
}
// @lc code=end

File: java/866.prime-palindrome.java
/*
 * @lc app=leetcode id=866 lang=java
 *
 * [866] Prime Palindrome
 */

// @lc code=start
class Solution {

    private static boolean isPrime(int N) {
        if (N <= 2) {
            return N == 2;
        }
        for (int i = (int)Math.sqrt(N); i >= 2; --i) {
            if (N % i == 0) {
                return false;
            }
        }
        return true;
    }

    public int primePalindrome(int N) {
        if (N >= 8 && N <= 11) {
            return 11;
        }
        for (int i = 0; i < 100000; ++i) {
            String s = Integer.toString(i);
            int x = Integer.parseInt(s.substring(0, s.length() - 1) + (new StringBuilder(s)).reverse().toString());
            if (x >= N && isPrime(x)) {
                return x;
            }
        }
        return -1;
    }
}
// @lc code=end

